zktrie part1: change storage proof from per step to per block

lispc opened the pull request:
Step-wise proof will be reconstructed in prover side.

Core design: when storage related opcodes are traced(eg: sstore, create2..), no storage proof are generated then, only "storage diff" are recorded. Later when sealing a block, the miner traverse all the "recorded accessed storage", and make proofs of them  for the state **before this block**. So for every storage slot,   there will be only one proof for it, corresponding to state before the block execution.


we add some methods to StateDb for this purpose. StateDb writes to trie only when a block is sealing, while during execution inside block, it only  keeps pending writes  in its inner buffer. So when we are making proofs, we can access state trie **before**  this block, since the pending changes are not committed yet.

lispc committed:
change storage proof from step-wise to block-wise. Step-wise proof will be reconstructed in prover side

lispc committed:
minor

lispc committed:
minor

noel2004 commented:
> Step-wise proof will be reconstructed in prover side.
> 
> Will this break current roller codes?

I don't think it would break roller's code because nothing in roller would touch the SMT proof fields, which would be purged from blockResult after this PR.

We can simply remove the "proof" fields which is defined in AccountProofWrapper/StorageProofWrapper@eth.rs in roller and the code still works.

mask-pp commented on core/types/l2trace.go:
```diff
@@ -1,34 +1,92 @@
 }
 
+// HexInt wrap big.Int for hex encoding
+type HexInt struct {
```
How about moving to the Common module？

noel2004 replied:
to hexutil module may be better?

mask-pp replied:
> to hexutil module may be better?

I agree, or just reuse hexutil.Big?

noel2004 replied:
That's good, i will just purged it because it is just used in the SMTTrace output (not included in current PR) and use hexutil.Big instead later

mask-pp commented on core/types/l2trace.go:
```diff
@@ -1,34 +1,92 @@
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (hi *HexInt) UnmarshalJSON(input []byte) error {
+	var s string
```
`s` is not been used.

noel2004 replied:
code has been purged

mask-pp commented on core/types/l2trace.go:
```diff
@@ -1,34 +1,92 @@
+// MarshalText implements encoding.TextMarshaler
+func (hi HexInt) MarshalJSON() ([]byte, error) {
+	if hi.Int == nil {
+		return json.Marshal("0x")
```
`return "0x"`

mask-pp committed:
Update worker.go

noel2004 committed:
purge hexInt

lispc commented:
@HAOYUatHZ will do a bit refactor here. So don't merge now.

lispc committed:
Merge branch 'zkrollup' into l2witness/opt-storage-proof

lispc committed:
Update l2trace.go

0xmountaintop committed:
Merge branch 'zkrollup' into l2witness/opt-storage-proof

0xmountaintop commented on core/vm/logger.go:
```diff
@@ -314,9 +305,68 @@ func (l *StructLogger) CaptureEnd(output []byte, gasUsed uint64, t time.Duration
 }
 
 func (l *StructLogger) CaptureEnter(typ OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	//the last logged op should be CALL/CREATE
```
should it be `CALL/CALLCODE/CREATE`?

lispc replied:
<del>there are five types of CALLxx. So either we list them all, or just put `CALL` here. I think a normal reader will understand all CALL(staticcall delegatecall call callcode..) should share same logic. They have no reason not share some procedure.</del>

not sure, wait for @noel2004 reply

noel2004 replied:
Yes, I simply missed CALLCODE ...
I would double check here and some related code

0xmountaintop commented on core/vm/logger_trace.go:
```diff
@@ -16,11 +14,11 @@ var (
 		CALL:         {traceToAddressCode, traceLastNAddressCode(1), traceCallerProof, traceLastNAddressProof(1)},
 		CALLCODE:     {traceToAddressCode, traceLastNAddressCode(1), traceCallerProof, traceLastNAddressProof(1)},
```
what's the different between `traceLastNAddressProof(1)` and doing it in `CaptureEnter`? If they are the same can we remove `traceLastNAddressProof` here?

then change `ExtraData` comments in `core/types/l2trace.go` to 
```
CALL | CALLCODE: [caller contract address’s account state, 
    callee contract address’s account state (before called)]
```

noel2004 replied:
The data created by traceLastNAddressProof is left for background compatible. I think we should use the account data captured in CaptureEnter, since the balance of called account may be updated and it would only be captured here.

0xmountaintop committed:
Refactor l2witness/opt-storage-proof (#112)

* rename GetStateData to GetLiveStateObject

* revert EvmTxTraces type

* rename GetLiveStateObject to GetLiveStateAccount

* fix typo

* some renamings

* format codes

* fix typo

* fix typos

* format codes

some reverts

some renamings

some renamings

format codes

* update comments

update comments

* update comments

update comments

update comments

* update comments

update comments

update comments

* rename

* rename

* update

* update comments
